middleware/dnstap/msg/msg.go

```diff
@@ -0,0 +1,134 @@
+// Package msg helps to build a dnstap Message.
+package msg
+
+import (
+	"time"
+
+	"github.com/coredns/coredns/request"
+
+	tap "github.com/dnstap/golang-dnstap"
+	"github.com/miekg/dns"
+	"github.com/pkg/errors"
+)
+
+// Data helps to build a dnstap Message.
+// It can be transformed into the actual Message using ToMsg.
+type Data struct {
+	Type        tap.Message_Type
+	Packed      []byte
+	SocketProto tap.SocketProtocol
+	SocketFam   tap.SocketFamily
+	Address     []byte
+	Port        uint32
+	TimeSec     uint64
+}
+
+func FromRequest(d *Data, r request.Request) error {
+	if err := networkFromWriter(d, r.W); err != nil {
+		return err
+	}
+	socket(d, &r)
+	return nil
+}
+
+func Pack(d *Data, m *dns.Msg) error {
+	packed, err := m.Pack()
+	if err != nil {
+		return err
+	}
+	d.Packed = packed
+	return nil
+}
+
+func Epoch(d *Data) {
+	d.TimeSec = uint64(time.Now().Unix())
+}
+
+// Transform the data into the actual message based on the type.
+func ToMsg(d *Data) *tap.Message {
+	m := tap.Message{
+		Type:           &d.Type,
+		SocketFamily:   &d.SocketFam,
+		SocketProtocol: &d.SocketProto,
+	}
+	switch *m.Type {
+	case tap.Message_CLIENT_QUERY,
+		tap.Message_RESOLVER_QUERY,
+		tap.Message_AUTH_QUERY,
+		tap.Message_FORWARDER_QUERY,
+		tap.Message_TOOL_QUERY:
+		// is query
+		m.QueryTimeSec = &d.TimeSec
+		m.QueryMessage = d.Packed
+	default:
```

Comment:
might want to make this explicit instead of `default` and leave `default` when nothing matches.